### PR TITLE
Adding mongodb based video player

### PR DESCRIPTION
**Pull request by cdondrup**
Inspired by the music player by @nilsbore. Uses the mjpeg server and opencv to play the video.

Issues: no sounds
Tried to look at pygame for this, but could not find anything obvious. Maybe @marc-hanheide has an idea.

This works for now and is used in the walking group entertainment interface.

### CODE REVIEW

File: mongodb_video_player/scripts/video_server.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            return
+        if not video.isOpened():
+            rospy.logwarn("Capture stream could not be opened")
+            return
+
+        ret, frame = video.read()
+        while self.play and ret and not rospy.is_shutdown():
+            img = self.bridge.cv2_to_imgmsg(frame, "bgr8")
+            self.video_pub.publish(img)
+            self.rate.sleep()
```
Comment (marc-hanheide) [inline mongodb_video_player/scripts/video_server.py]: so, this assumes always be 25Hz videos. Quite a bold assumption. Can the rate be read from the video file?

Comment (cdondrup) [inline mongodb_video_player/scripts/video_server.py]: Valid point. I'll have a look but I guess so.
Don't merge yet.

Comment (marc-hanheide) [inline mongodb_video_player/scripts/video_server.py]: `fps = video.get(cv.CV_CAP_PROP_FPS)`

**Comment by cdondrup** [comment]
Reading fps from video and setting appropriate sleep rate.

File: mongodb_video_player/scripts/video_server.py (hunk 1 above)
Comment (cdondrup) [inline mongodb_video_player/scripts/video_server.py]: Exactly :)

**Comment by marc-hanheide** [comment]
for the audio: No, I don't have any idea, other than going full-scale with using ffmpeg library etc, But synchronicity will not be possible with the current approach. The alternative would be to not use the mjpeg canvas, but to really integrate an HTML5 player:

```
<video width="356" height="200" controls poster="full/http/link/to/image/file.png"  >

<source src="full/link/to/http/mp4/video/file.mp4" type="video/mp4" />

<source src="full/link/to/http/ogv/video/file.ogv" type="video/ogg" />

<source src="full/link/to/http/webm/video/file.wbem" type="video/webm" />

<em>Sorry, your browser doesn't support HTML5 video.</em>

</video>
```

Then we would have to be able to serve the files via HTTP. That actually looks better, I believe...

**Comment by cdondrup** [comment]
OK, that would be good. My only concern is that I have to hard code the path to the video files if I understand that correctly. I can use the media server to retrieve them but the html has to contain the path where I put them. Can I use absolute paths in there? I don't think so... how would I tell it where to look for the video?

**Comment by cdondrup** [comment]
So basically, my problem is the same as with the images. The html files are all installed to `/opt/ros/indigo` and the webserver is only able to display things relative to that directory. That would mean I have to put the videos and images into that directory which I can't due to permissions. The mjpeg server allows me to show images published to a topic so I can just retrieve everything from the mongodb, put it somewhere I have permissions and read it.

Maybe I got something wrong and that is easy to fix, but afaik I cannot tell the html5 player where to look for the video, because it would only be able to find it in a directory where I don't have write permissions.

**Comment by cdondrup** [comment]
If you don't mind @marc-hanheide I merge this because it does build, install, and works for now, so I can continue working on the overall system. This can be fixed at a later point.

**Comment by marc-hanheide** [comment]
Go, merge it, but I think this calls for a change in strands_webserver as we already discussed in #59.

**Comment by cburbridge** [comment]
The media_server package already serves the video files over HTTP, so no need to have another webserver to serve them or write them to a file in addition to the file already in the mongodb...

**Comment by cdondrup** [comment]
@cburbridge How would I show a video on a webpage then using the media server?
I'm rather bad at all these web things, sorry.

**Comment by marc-hanheide** [comment]
check https://github.com/strands-project/strands_ui/blob/hydro-devel/mongodb_media_server/scripts/server.py#L43

seems you can simply use

```
<video width="356" height="200" >
<source src="http://MEDIAHOST:8027/get_media/<ID>" type="video/mp4" />
<em>Sorry, your browser doesn't support HTML5 video.</em>
</video>
```

**Comment by marc-hanheide** [comment]
Sorry to render your video_server void ;-)

**Comment by cdondrup** [comment]
Even better. Maybe I don't need the image server hack then either.

**Comment by cburbridge** [comment]
Exactly as @marc-hanheide says. At the moment I only provided a Python API to query the list of available 'sets' and get the ID of files in the set (see README), but an http based API to allow javascript based queries is (low) on my todo list.

**Pull request closed without merging**